### PR TITLE
qordered attn draft

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+REGISTER_KERNEL_TYPED(int8_t)
+
+QOrderedAttention::QOrderedAttention(const OpKernelInfo& info) : CudaKernel(info), AttentionBase(info) {
+  order_input_ = GetCublasLtOrderAttr(info, "order_input");
+  order_weight_ = GetCublasLtOrderAttr(info, "order_weight");
+  order_bias_ = GetCublasLtOrderAttr(info, "order_bias");
+  order_output_ = GetCublasLtOrderAttr(info, "order_output");
+  //TODO: any checks?
+  // bugbug
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:36]: question: do we need to check the order? would there be other order combinations in the future in onnx model?

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  const Tensor* scale_bias = context->Input<Tensor>(5);
+  const Tensor* scale_output = context->Input<Tensor>(7);
+
+  const float* scale_input_data = scale_input->template Data<float>();
+  const float* scale_weights_data = scale_weights->template Data<float>();
+  const float* scale_bias_data = scale_bias->template Data<float>();
+  const float* scale_output_data = scale_output->template Data<float>();
+
+  // bugbug
+  ORT_UNUSED_PARAMETER(scale_output_data);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:60]: question: not sure what's the last scale in this op and how to use it

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  ORT_RETURN_IF_ERROR(
+    Reorder(cublasLt, stream, gsl::narrow_cast<int>(1), m, n, CUDA_R_8I,
+            gemm_buffer_quantized_col32.get(), (cublasLtOrder_t)2, gemm_buffer_quantized_row.get(), (cublasLtOrder_t)0));
+
+  using CudaT = ToCudaType<MLFloat16>::MappedType;
+  constexpr size_t element_size = sizeof(MLFloat16);
+
+  // dequantize back to fp16
+  auto gemm_buffer = GetScratchBuffer<int8_t>(batch_size * sequence_length * 3 * hidden_size * element_size); //row, fp16
+  const CudaT scale = (CudaT)1.0f; // bugbug
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:118]: question: dequantize scale or bias?

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+          nullptr,
+          nullptr)) {
+    // Get last error to reset it to cudaSuccess.
+    CUDA_CALL(cudaGetLastError());
+    return Status(common::ONNXRUNTIME, common::FAIL);
+  }
+
+  //quantize to int8
+  auto output_buffer_quantized = GetScratchBuffer<int8_t>(batch_size * sequence_length * hidden_size * 1); //row, int8
+  const CudaT scale2 = (CudaT)1.0f; // bugbug
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:153]: question: scale/bias for quantize

File: onnxruntime/contrib_ops/cuda/quantization/qorder_common.cc [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+                                        scale, A, desc_A, B, desc_B,
+                                        &beta, C, desc_C, C, desc_C,
+                                        &algo, nullptr, 0,  // algo, workspace, workspace_size
+                                        stream));
+
+  return Status::OK();
+}
+
+// Matmul descriptor must specify CUBLAS_OP_T on matrix B and CUBLAS_OP_N (default) on matrix A and C
+Status QOrdered_Gemm(cublasLtHandle_t cublasLt_handle, cudaStream_t stream,
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_common.cc:195]: Added a gemm wrapper. TODO: refactor

File: onnxruntime/core/graph/contrib_ops/quantization_defs.cc [hunk 6]
```diff
@@ -988,5 +988,171 @@ Wwhere the function `Sigmoid(x) = 1 / (1 + exp(-x))` )DOC";
...
+      .Input(1, "scale_input", "scale of the input", "S")
+      .Input(2, "weight", "2D input tensor with shape (input_hidden_size, 3 * hidden_size), where hidden_size = num_heads * head_size", "Q")
+      .Input(3, "scale_weight", "scale of the weight", "S")
+      .Input(4, "bias", "1D input tensor with shape (3 * hidden_size)", "Q")
+      .Input(5, "scale_bias", "scale of the bias", "S")
+      .Input(6, "mask_index",
+             "Attention mask with shape (batch_size, 1, max_sequence_length, max_sequence_length), (batch_size, past_sequence_length + sequence_length)"
+             "or (batch_size, sequence_length, past_sequence_length + sequence_length), or index with shape (batch_size) or (2 * batch_size).",
+             "G", OpSchema::Optional)
+      .Input(7, "scale_output", "scale of the output", "S")
```
Comment (wangyems) [inline onnxruntime/core/graph/contrib_ops/quantization_defs.cc:1147]: question: not sure if this is the scale for output

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+#include "qorder_common.h"
+
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+using namespace onnxruntime::cuda;
+
+class QOrderedAttention final : public CudaKernel, public AttentionBase {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.h:18]: TODO: add UT

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc (hunk 2 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:60]: You could just scale_bias_data = scale_bias_data / scale_output_data;

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc (hunk 3 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:118]: TODO: add gemm output scale

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc (hunk 4 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:153]: output scale

File: onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  const float scale_alpha = *scale_input_data * (*scale_weights_data);
+  ORT_RETURN_IF_ERROR(
+    QOrdered_Gemm(cublasLt, stream, (int)1, m, n, k, &scale_alpha,
+                  reinterpret_cast<const int8_t*>(input->template Data<int8_t>()),
+                  reinterpret_cast<const int8_t*>(weights->template Data<int8_t>()),
+                  &scale_beta, reinterpret_cast<int8_t*>(gemm_buffer_quantized_col32.get()),
+                  (cublasLtOrder_t)order_input_, (cublasLtOrder_t)order_weight_, (cublasLtOrder_t)order_bias_, device_prop));
+
+  // reorder to row major
+  auto gemm_buffer_quantized_row = GetScratchBuffer<int8_t>(batch_size * sequence_length * 3 * hidden_size * 1);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/quantization/qorder_attention.cc:108]: rebase with qorder2 and use one buffer later

**Pull request closed without merging**